File: android/app/src/main/kotlin/com/aispotlight/android/chat/RemoteFileJanitor.kt
package com.aispotlight.android.chat

import android.content.Context
import com.aispotlight.android.core.Diagnostics
import com.aispotlight.android.core.ProviderException
import com.aispotlight.android.core.ProviderID
import com.aispotlight.android.providers.OpenAIFilesService
import com.aispotlight.android.settings.ApiKeyStore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.json.JSONArray
import org.json.JSONObject

/**
 * Best-effort deletion of provider-side document copies. Every place that
 * lets an attachment go enqueues the remote ids here; the queue is persisted
 * so an offline deletion is retried at the next launch. Server-side expiry
 * (`expires_after` at upload) is the backstop. Port of the Mac janitor.
 */
object RemoteFileJanitor {
    private const val PREFS = "files.janitor"
    private const val KEY_QUEUE = "queue"
    private const val MAX_ENTRIES = 200

    private lateinit var appContext: Context
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val drainMutex = Mutex()

    fun init(context: Context) {
        appContext = context.applicationContext
    }

    /** Adds ids (deduplicated) and kicks a drain. */
    fun enqueue(provider: ProviderID, fileIds: Collection<String>) {
        val fresh = fileIds.filter { it.isNotEmpty() }
        if (fresh.isEmpty() || !::appContext.isInitialized) return
        synchronized(this) {
            val entries = load().toMutableList()
            for (id in fresh) {
                val entry = provider.id to id
                if (entry !in entries) entries.add(entry)
            }
            while (entries.size > MAX_ENTRIES) entries.removeAt(0)
            save(entries)
            Diagnostics.log("files", "janitor enqueue count=${fresh.size} queued=${entries.size}")
        }
        drainSoon()
    }

    fun drainSoon() {
        if (!::appContext.isInitialized) return
        scope.launch { drain() }
    }

    private suspend fun drain() = drainMutex.withLock {
        val entries = synchronized(this) { load() }
        if (entries.isEmpty()) return@withLock
        val remaining = mutableListOf<Pair<String, String>>()
        for ((providerId, fileId) in entries) {
            val provider = ProviderID.fromId(providerId)
            if (provider != ProviderID.OPENAI) continue // nothing to call: drop
            val apiKey = ApiKeyStore.key(provider)
            if (apiKey == null) { remaining.add(providerId to fileId); continue }
            try {
                OpenAIFilesService.delete(fileId, apiKey)
                Diagnostics.log("files", "delete id=$fileId status=ok")
            } catch (e: ProviderException) {
                val message = e.message ?: ""
                // Client errors won't heal on retry (bad id, revoked key).
                if (Regex("HTTP 4\\d\\d").containsMatchIn(message)) {
                    Diagnostics.log("files", "delete id=$fileId dropped (${message.take(80)})")
                } else {
                    Diagnostics.log("files", "delete id=$fileId failed — kept (${message.take(80)})")
                    remaining.add(providerId to fileId)
                }
            } catch (e: Exception) {
                Diagnostics.log("files", "delete id=$fileId failed — kept (${e.message?.take(80)})")
                remaining.add(providerId to fileId)
            }
        }
        synchronized(this) {
            // Entries added while draining stay; only the processed ones go.
            val processed = entries.map { it.second }.toSet() - remaining.map { it.second }.toSet()
            save(load().filter { it.second !in processed })
        }
    }

    private fun load(): List<Pair<String, String>> {
        val raw = appContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE).getString(KEY_QUEUE, null)
            ?: return emptyList()
        return try {
            val array = JSONArray(raw)
            (0 until array.length()).mapNotNull { index ->
                val item = array.optJSONObject(index) ?: return@mapNotNull null
                item.optString("provider") to item.optString("fileId")
            }.filter { it.second.isNotEmpty() }
        } catch (_: Exception) {
            emptyList()
        }
    }

    private fun save(entries: List<Pair<String, String>>) {
        val prefs = appContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        if (entries.isEmpty()) {
            prefs.edit().remove(KEY_QUEUE).apply()
            return
        }
        val array = JSONArray()
        for ((provider, fileId) in entries) {
            array.put(JSONObject().put("provider", provider).put("fileId", fileId))
        }
        prefs.edit().putString(KEY_QUEUE, array.toString()).apply()
    }
}

File: Cuate/Providers/RemoteFileJanitor.swift
import Foundation

/// Best-effort deletion of provider-side document copies. Every place that
/// lets an attachment go (message deleted, new chat, preset deleted, the
/// 15-day prune) enqueues the remote ids here; the queue is persisted so an
/// offline deletion is retried on the next launch. Server-side expiry
/// (`expires_after` at upload) is the backstop when this never succeeds.
nonisolated enum RemoteFileJanitor {
    struct Entry: Codable, Equatable {
        let provider: String
        let fileID: String
    }

    private static let defaultsKey = "files.janitor.queue"
    /// Bounded: a runaway caller can't grow the defaults without limit.
    private static let maxEntries = 200
    private static let queue = DispatchQueue(label: "Cuate.RemoteFileJanitor", qos: .utility)
    private static var draining = false

    /// Adds ids to the queue (deduplicated) and kicks a drain.
    static func enqueue(provider: ProviderID, fileIDs: [String]) {
        let fresh = fileIDs.filter { !$0.isEmpty }
        guard !fresh.isEmpty else { return }
        queue.async {
            var entries = load()
            for id in fresh {
                let entry = Entry(provider: provider.rawValue, fileID: id)
                if !entries.contains(entry) { entries.append(entry) }
            }
            if entries.count > maxEntries {
                entries.removeFirst(entries.count - maxEntries)
            }
            save(entries)
            Diagnostics.log("files", "janitor enqueue count=\(fresh.count) queued=\(entries.count)")
        }
        drainSoon()
    }

    /// Runs the queue in the background; safe to call often (one drain at a time).
    static func drainSoon() {
        queue.async {
            guard !draining, !load().isEmpty else { return }
            draining = true
            Task.detached(priority: .utility) {
                await drain()
                queue.async { draining = false }
            }
        }
    }

    private static func drain() async {
        await APIKeyStore.warmIfNeeded()
        let entries = queue.sync { load() }
        var remaining: [Entry] = []
        for entry in entries {
            guard let provider = ProviderID(rawValue: entry.provider), provider == .openai else {
                continue // unknown provider: nothing to call, drop the entry
            }
            guard let apiKey = APIKeyStore.key(for: provider) else {
                remaining.append(entry) // no key right now — keep for later
                continue
            }
            do {
                try await OpenAIFilesService.delete(fileID: entry.fileID, apiKey: apiKey)
                Diagnostics.log("files", "delete id=\(entry.fileID) status=ok")
            } catch ProviderError.http(let status, let message) where (400..<500).contains(status) {
                // Client errors won't heal on retry (bad id, revoked key).
                Diagnostics.log("files", "delete id=\(entry.fileID) status=\(status) dropped (\(message.prefix(80)))")
            } catch {
                Diagnostics.log("files", "delete id=\(entry.fileID) failed — kept (\(error.localizedDescription.prefix(80)))")
                remaining.append(entry)
            }
        }
        queue.sync {
            // Entries added while draining stay; only the processed ones go.
            let current = load()
            let processed = Set(entries.map { $0.fileID }).subtracting(remaining.map { $0.fileID })
            save(current.filter { !processed.contains($0.fileID) })
        }
    }

    // MARK: - Storage (call on `queue` only)

    private static func load() -> [Entry] {
        guard let data = UserDefaults.standard.data(forKey: defaultsKey),
              let entries = try? JSONDecoder().decode([Entry].self, from: data) else { return [] }
        return entries
    }

    private static func save(_ entries: [Entry]) {
        if entries.isEmpty {
            UserDefaults.standard.removeObject(forKey: defaultsKey)
        } else if let data = try? JSONEncoder().encode(entries) {
            UserDefaults.standard.set(data, forKey: defaultsKey)
        }
    }
}
